### PR TITLE
NEW: @W-18097268@: Diagnostics now respond to text document changes and mark diagnostics stale if needed.

### PR DESCRIPTION
**Pull request by stephen-carter-at-sf**

### CODE REVIEW

File: src/test/unit/lib/diagnostics-handleTextDocumentChangeEvent.test.ts
Comment (stephen-carter-at-sf) [inline src/test/unit/lib/diagnostics-handleTextDocumentChangeEvent.test.ts:1]: This test file is large - so github is probably going to show it collapsed. But it is the heart of this change to lock in all the edge cases.

File: src/lib/agentforce/agentforce-code-action-provider.ts [hunk 1]
```diff
@@ -28,7 +28,7 @@ export class AgentforceCodeActionProvider implements vscode.CodeActionProvider {
         const codeActions: vscode.CodeAction[] = [];
         const filteredDiagnostics: CodeAnalyzerDiagnostic[] = context.diagnostics
             .filter(d => d instanceof CodeAnalyzerDiagnostic)
-            .filter(d => range.contains(d.range) && A4D_SUPPORTED_RULES.has(d.violation.rule));
+            .filter(d => !d.isStale() && range.contains(d.range) && A4D_SUPPORTED_RULES.has(d.violation.rule));
```
Comment (randi274) [inline src/lib/agentforce/agentforce-code-action-provider.ts:31]: This stale behavior change is good, as it won't show as a quick fix option anymore. It's probably worth calling out in the docs too @jshackell-sfdc

File: src/lib/diagnostics.ts [hunk 2]
```diff
@@ -188,3 +220,109 @@ function adjustToZeroBased(value: number): number {
...
 }
+
+
+/**
+ * Algorithm to adjust a diagnostic's range (or discard it by return null) based on the text document change event.
+ * This algorithm needs to be very fast since it literally runs on every stroke of a key within the editor window.
+ */
+function adjustDiagnosticToChange(diag: CodeAnalyzerDiagnostic, change: vscode.TextDocumentContentChangeEvent,
+                                  replacementLines: string[]): CodeAnalyzerDiagnostic | null {
+    // Key: .  single line (i.e. no '\n' characters)
```
Comment (randi274) [inline src/lib/diagnostics.ts:231]: Helpful 👍

Comment (stephen-carter-at-sf) [reply src/lib/diagnostics.ts:231]: Yeah I needed a way of having short hand to cover all the cases... and to somehow visualize what i'm talking about. This is more for me and the way my brain works.

File: src/test/unit/lib/diagnostics-handleTextDocumentChangeEvent.test.ts [hunk 3]
```diff
@@ -0,0 +1,785 @@
...
+        //      {.}->{.}  single line change range replaced by single line
+        //      {.}->{_}  single line change range replaced by >1 lines
+        //      {_}->{.}  multi line change range replaced by single line
+        //      {_}->{_}  multi line change range replaced by >1 lines
+        //      [.]       diagnostic range containing a 1 line
+        //      [_]       diagnostic range containing >1 lines
+
+
+        // _{.}->{.}_[.]_
+        it('1 line change range is before a 1 line diagnostic range (on separate lines). Replacement is 1 line', () => {
```
Comment (randi274) [inline src/test/unit/lib/diagnostics-handleTextDocumentChangeEvent.test.ts:145]: Small suggestion here to map each block of tests to the "cases" you've defined well in `diagnostics.ts`, to make it easier to what tests are responsible for what (and future changes anywhere in there will be easier to touch here too). The individual comments are good, but the case mapping would be helpful too.
Screenshot 2025-03-31 at 11 27 30 AM

**Review by randi274 — DISMISSED** [review]
A few suggestions, but looking really good overall - great work and comments on the `adjustDiagnosticToChange` algorithm!

File: src/test/unit/lib/diagnostics-handleTextDocumentChangeEvent.test.ts (hunk 3 above)
Comment (stephen-carter-at-sf) [reply src/test/unit/lib/diagnostics-handleTextDocumentChangeEvent.test.ts:145]: With test driven development, I wrote the tests first and did a ton of refactoring afterward  and it just so happen that the source code looks a lot like the test code.. but I don't think I want to tie the source code's comments to be 100% coupled to the test codes comments and vice verca. We should have the liberty to completely replace the source code and not touch the tests. So although I understand where you are coming from, I don't know if it's necessary/wanted.

File: src/lib/diagnostics.ts [hunk 4]
```diff
@@ -29,6 +29,8 @@ export type Violation = {
     resources: string[];
 }
 
+const STALE_PREFIX = '(STALE: Code has changed. Re-run needed.)\n';
```
[the remark this answers is not in the document]
Comment (stephen-carter-at-sf) [reply src/lib/diagnostics.ts]: Yes. Thanks. Done.

**Review by randi274 — APPROVED** [review]
